File: Backend/Management Coding Tasks/tasksManager.vb
Public Class tasksManagerClass
    Public Const TO_START As Boolean = -1
    Public Const BUSY As Boolean = 1
    Public Const FINISHED As Boolean = 0

    Public Event tasksCompleted(sender As Object)


    Private tasks As Dictionary(Of String, Integer)
    Private WithEvents timer As Timer


    Public Sub New()
        tasks = New Dictionary(Of String, Integer)
    End Sub

    Public Sub registerTask(name As String, status As Integer)
        tasks.Add(name, status)
    End Sub

    Public Sub startListening()
        timer = New Timer
        timer.Interval = 100
        timer.Start()

    End Sub

    Public Sub unload()
        If timer IsNot Nothing Then
            timer.Stop()
        End If
    End Sub
    Private Sub timer_tick(ByVal sender As Object, ByVal e As EventArgs) Handles timer.Tick
        If getTasksStatus.Equals(FINISHED) Then
            RaiseEvent tasksCompleted(Me)
            timer.Stop()
        End If
    End Sub

    Public Sub setStatus(name As String, status As Integer)
        tasks(name) = status
    End Sub

    Public Function getTasksStatus() As Integer
        For i = 0 To tasks.Count - 1
            If tasks.ElementAt(i).Value.Equals(TO_START) Or tasks.ElementAt(i).Value.Equals(BUSY) Then
                Return BUSY
            End If
        Next i
        Return FINISHED
    End Function
End Class

File: Backend/Management Coding Tasks/Management Coding Tasks/tasksManager.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace AeonLabs.tasksManager
{
    public class tasksManagerClass
    {
        public const int TO_START = -1;
        public const int BUSY = 1;
        public const int FINISHED = 0;

        public event tasksCompletedEventHandler tasksCompleted;

        public delegate void tasksCompletedEventHandler(object sender);

        private Dictionary<string, int> tasks;
        private System.Timers.Timer atimer = new System.Timers.Timer();
        private bool taskEnded = false;

        public tasksManagerClass()
        {
            tasks = new Dictionary<string, int>();
        }

        public void registerTask(string name, int status)
        {
            tasks.Add(name, status);
        }

        public void startListening()
        {
            taskEnded = false;
            atimer.AutoReset = true;
            atimer.Interval = 100;
            atimer.Elapsed += tick;
            atimer.Start();
        }

        public void unload()
        {
            if (atimer is object)
            {
                atimer.Stop();
            }
        }

        private void tick(object sender, EventArgs e)
        {
            if (getTasksStatus().Equals(FINISHED) & !taskEnded)
            {
                taskEnded = true;
                tasksCompleted?.Invoke(this);
                atimer.Stop();
            }
        }

        public void setStatus(string name, int status)
        {
            tasks[name] = status;
        }

        public int getTasksStatus()
        {
            for (int i = 0, loopTo = tasks.Count - 1; i <= loopTo; i++)
            {
                if (tasks.ElementAt(i).Value.Equals(TO_START) | tasks.ElementAt(i).Value.Equals(BUSY))
                {
                    return BUSY;
                }
            }

            return FINISHED;
        }
    }
}